src/date.c

```diff
@@ -1,18 +1,981 @@
+    /* check results */
+    int valid = 0;
+    
+    tp.month--;
+    valid = tp.day >=1 && tp.day <= 31;
+    valid &= tp.month >=0 && tp.month <= 11;
+    valid &= tp.hour >= 0 && tp.hour <= 23;
+    valid &= tp.min >= 0 && tp.min <= 59;
+    valid &= tp.sec >= 0 && tp.sec <= 59;
+    
+    if(tz != NO_TZ && tz > 12) {
+      tz /= 100;
+    }
+    
+    valid &= (abs(tz) <= 12 || tz == NO_TZ);
+    
+    if(valid) {
+      *time = d_gmktime(&tp);
+
+      if(tz != NO_TZ) {
+        /* timezone specified, use it */
+        *time -= tz * msPerHour;
+      } else if(tz != 0 ) {
+        /* assuming local timezone and moving back to UTC */
+        *time = ecma_UTC(*time);
+      }
+    }
+  }
+
+  return !isnan(*time);
+}
+
+typedef int (*ftostring)(const struct timeparts*, char*, int);
+
+static val_t d_tostring(struct v7 *v7, val_t obj, fbreaktime breaktimefunc,
+                                        ftostring tostringfunc, int addtz) {
+  struct timeparts tp;
+  int len;
+  char buf[100];
+  etime_t time;
+  
+  time = v7_to_double(d_trytogetobjforstring(v7, obj));
+  
+  breaktimefunc(&time, &tp);
+  len = tostringfunc(&tp, buf, addtz);
+  
+  return v7_create_string(v7, buf, len, 1);
+}
+
+struct d_locale {
+  char locale[50];
+};
+
+static void d_getcurrentlocale(struct d_locale *loc) {
+  strcpy(setlocale(LC_TIME, 0), loc->locale);
+}
+
+static void d_setlocale(const struct d_locale *loc) {
+  setlocale(LC_TIME, loc? loc->locale: "");
+}
+
+/* TODO(alashkin): check portability */
+static val_t d_tolocalestr(struct v7 *v7, val_t obj, const char *frm) {
+  char buf[250];
+  size_t len;
+  struct tm t;
+  etime_t time;
+  struct d_locale prev_locale;
+  struct timeparts tp;
+
+  time = v7_to_double(d_trytogetobjforstring(v7, obj));
+  
+  d_getcurrentlocale(&prev_locale);
+  d_setlocale(0);
+
+  d_localtime(&time, &tp);
+
+  memset(&t, 0, sizeof(t));
+  t.tm_year = tp.year - 1900;
+  t.tm_mon = tp.month;
+  t.tm_mday = tp.day;
+  t.tm_hour = tp.hour;
+  t.tm_min = tp.min;
+  t.tm_sec = tp.sec;
+  t.tm_wday = tp.dayofweek;
+
+  len = strftime(buf, sizeof(buf), frm, &t);
+  
+  d_setlocale(&prev_locale);
+  
+  return v7_create_string(v7, buf, len, 1);
+}
+
+static int d_tptodatestr(const struct timeparts *tp, char *buf, int addtz) {
+  (void)addtz;
+
+  return sprintf(buf, "%s %s %02d %d", wday_name[tp->dayofweek],
+                                    mon_name[tp->month], tp->day, tp->year);
+}
+
+static int d_tptotimestr(const struct timeparts *tp, char *buf, int addtz) {
+  int len;
+
+  len = sprintf(buf, "%02d:%02d:%02d GMT", tp->hour, tp->min, tp->sec);
+  
+  if(addtz && g_gmtoffms != 0) {
+    len = sprintf(buf+len, "%c%02d00 (%s)", g_gmtoffms > 0? '-':'+',
+                                abs((int)g_gmtoffms/msPerHour), d_gettzname());
+  }
+  
+  return (int)strlen(buf);
+}
+
+static int d_tptostr(const struct timeparts *tp, char *buf, int addtz) {
+  int len = d_tptodatestr(tp, buf, addtz);
+  *(buf + len) = ' ';
```

Comment:
For me - it is not working with array, it is pointer operation
>> int len = d_tptodatestr(tp, buf, addtz);
>> *(buf + len) = ' ';
>>return d_tptotimestr(tp, buf + len + 1, addtz) + len + 1;

I use (buf + len) as a start point - and as result I wrote *(buf+len) instead of buf[len]